File: in_Python/1160 Find Words That Can Be Formed by Characters.py
class Solution(object):
    def countCharacters(self, words, chars):
        """
        :type words: List[str]
        :type chars: str
        :rtype: int
        """
        res = 0
        dict_target = {}
        for c in list(chars):
            if c not in dict_target:
                dict_target[c] = 1
            else:
                dict_target[c] += 1
                                
        for w in words:
            dict = {}
            flag = False
            for i in list(w):
                if i not in dict_target:
                    flag = True
                    break
                else:
                    if i not in dict:
                        dict[i] = 1
                    else:
                        dict[i] += 1
            if flag:
                continue
            for k in dict:
                if dict[k] > dict_target[k]:
                    flag = True
                    break
            if flag:
                continue
            else:
                res += len(w)
        return res
                
        

File: in_Java/1160 Find Words That Can Be Formed by Characters.java
class Solution {
    public int countCharacters(String[] words, String chars) {
        HashMap<Character,Integer> hm = new HashMap<>();
        for(char c : chars.toCharArray()){
            if(!hm.containsKey(c)) hm.put(c,1);
            else hm.put(c, hm.get(c)+1);
        }
        int res = 0;
        for(String s : words){
            boolean f = true;
            HashMap<Character,Integer> t = new HashMap<>();
            for(char c : s.toCharArray()){
                if(!t.containsKey(c)) t.put(c,1);
                else t.put(c, t.get(c)+1);
            }
            for(char k : t.keySet()){
                if(!hm.containsKey(k)) {f = false;break;}
                else{
                    if(hm.get(k) < t.get(k)) {f = false;break;}
                }
            }
            if(f) res += s.length();
        }
        return res;
    }
}